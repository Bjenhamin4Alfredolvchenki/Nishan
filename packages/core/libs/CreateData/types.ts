import { ICache } from '@nishans/cache';
import { IOperation, TTextFormat } from '@nishans/types';
import { Logger } from '../../types';

export interface ParentCollectionData {
	parent_collection_id: string;
	name: TTextFormat;
	token: string;
	logger?: Logger;
	stack: IOperation[];
	cache: ICache;
	parent_relation_schema_unit_id: string;
}
