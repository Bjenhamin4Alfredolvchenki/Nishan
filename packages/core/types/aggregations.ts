import { TSchemaUnitType, IViewAggregationsAggregators } from '@nishans/types';

interface ViewAggregationsCreateGenericParam<T extends TSchemaUnitType> {
	schema_type: T;
	name: string;
	aggregator: IViewAggregationsAggregators[T];
}

export type UserViewAggregationsCreateParams =
	| ViewAggregationsCreateGenericParam<'text'>
	| ViewAggregationsCreateGenericParam<'title'>
	| ViewAggregationsCreateGenericParam<'number'>
	| ViewAggregationsCreateGenericParam<'select'>
	| ViewAggregationsCreateGenericParam<'multi_select'>
	| ViewAggregationsCreateGenericParam<'date'>
	| ViewAggregationsCreateGenericParam<'person'>
	| ViewAggregationsCreateGenericParam<'file'>
	| ViewAggregationsCreateGenericParam<'checkbox'>
	| ViewAggregationsCreateGenericParam<'url'>
	| ViewAggregationsCreateGenericParam<'email'>
	| ViewAggregationsCreateGenericParam<'phone_number'>
	| ViewAggregationsCreateGenericParam<'formula'>
	| ViewAggregationsCreateGenericParam<'relation'>
	| ViewAggregationsCreateGenericParam<'rollup'>
	| ViewAggregationsCreateGenericParam<'created_time'>
	| ViewAggregationsCreateGenericParam<'created_by'>
	| ViewAggregationsCreateGenericParam<'last_edited_time'>
	| ViewAggregationsCreateGenericParam<'last_edited_by'>;
