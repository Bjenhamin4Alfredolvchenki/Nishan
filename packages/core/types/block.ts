import { IAudio, IBreadcrumb, IBulletedList, ICallout, ICode, ICodepen, ICollection, IDivider, IDrive, IEmbed, IEquation, IFactory, IFigma, IFile, IGist, IHeader, IImage, IMaps, INotionUser, INumberedList, IPage, IQuote, ISpace, ISpaceView, ISubHeader, ISubSubHeader, IText, ITOC, ITodo, IToggle, ITweet, IUserSettingsSettings, IVideo, IWebBookmark, TBlockType, TTextFormat } from "@nishans/types";
import { TSchemaUnitInput } from "./schema";
import { ElementType } from "./utils";
import { TViewCreateInput } from "./view";

interface IInput {
  id?: string,
  type: TBlockType
}

export interface ICollectionBlockInput {
  id?: string,
  views: TViewCreateInput[],
  schema: TSchemaUnitInput[],
  name: TTextFormat,
  icon?: string;
  cover?: string;
  rows?: IPageCreateInput[],
  collection_id?:string
}

export interface ICollectionViewInput extends ICollectionBlockInput {
  type: "collection_view",
}

export interface ICollectionViewPageInput extends ICollectionBlockInput {
  type: "collection_view_page",
  isPrivate?: boolean
}

export interface ILinkedDBInput extends IInput {
  type: "linked_db",
  collection_id: string,
  properties?: Record<string, unknown>,
  format?: Record<string, unknown>,
  views: TViewCreateInput[],
}

export type TCollectionBlockInput = ICollectionViewInput | ICollectionViewPageInput | ILinkedDBInput;

// -----------------

// Media IBlock Input

export interface IVideoInput extends IInput {
  type: 'video',
  properties: IVideo["properties"],
  format?: IVideo["format"],
}

export interface IImageInput extends IInput {
  type: 'image',
  properties: IImage["properties"],
  format?: IImage["format"],
}

export interface IAudioInput extends IInput {
  type: 'audio',
  properties: IAudio["properties"],
  format?: IAudio["format"],
}

export interface IWebBookmarkInput extends IInput {
  type: 'bookmark',
  properties: IWebBookmark["properties"],
  format?: IWebBookmark["format"]
}

// Basic block input
export interface ICodeInput extends IInput {
  type: 'code',
  properties: ICode["properties"]
  format?: ICode["format"]
}

export interface IFileInput extends IInput {
  type: 'file',
  properties: IFile["properties"]
  format?: IFile["format"]
}

export type TMediaBlockInput = IVideoInput | IImageInput | IAudioInput | IWebBookmarkInput | ICodeInput | IFileInput;

// Basic IBlock Input

export interface IColumnListInput extends IInput {
  type: "column_list",
  contents: TBlockInput[]
}

export interface IPageCreateInput extends IInput {
  type: 'page',
  properties: IPage["properties"],
  format?: Partial<IPage["format"]>,
  isPrivate?: boolean,
  contents?: TBlockInput[],
  is_template?: boolean
}

export interface IPageUpdateInput {
  properties: Partial<IPage["properties"]>,
  format?: Partial<IPage["format"]>,
  isPrivate?: boolean,
  type: "page"
}

export interface ICollectionViewPageUpdateInput{
  type: "collection_view_page";
  cover?:string;
  icon?:string;
  isPrivate?:boolean
  title?:string[][]
}

export interface ITextInput extends IInput {
  properties: IText["properties"],
  format?: IText["format"],
  type: 'text'
}

export interface IHeaderInput extends IInput {
  properties: IHeader["properties"],
  format?: IHeader["format"],
  type: 'header'
}

export interface ISubHeaderInput extends IInput {
  properties: ISubHeader["properties"],
  format?: ISubHeader["format"],
  type: 'sub_header'
}

export interface ISubSubHeaderInput extends IInput {
  properties: ISubSubHeader["properties"],
  format?: ISubSubHeader["format"],
  type: 'sub_sub_header'
}

export interface INumberedListInput extends IInput {
  properties: INumberedList["properties"],
  format?: INumberedList["format"],
  type: 'numbered_list'
}

export interface IBulletedListInput extends IInput {
  properties: IBulletedList["properties"],
  format?: IBulletedList["format"],
  type: 'bulleted_list'
}

export interface IToggleInput extends IInput {
  properties: IToggle["properties"],
  format?: IToggle["format"],
  type: 'toggle'
}

export interface IQuoteInput extends IInput {
  properties: IQuote["properties"],
  format?: IQuote["format"],
  type: 'quote'
}

export interface ILinkToPageInput {
  type: "link_to_page",
  page_id: string,
}

export interface IDividerInput extends IInput {
  type: 'divider',
  properties?: IDivider["properties"],
  format?: IDivider["format"]
}

export interface ICalloutInput extends IInput{
  type: 'callout',
  format?: ICallout["format"]
  properties?: ICallout["properties"]
}

export interface ITodoInput extends IInput {
  type: 'to_do',
  properties: ITodo["properties"],
  format?: ITodo["format"]

}
// ? TD:2:M Add td for TCollectionBlockInput

export type TBasicBlockInput = ILinkToPageInput | IPageCreateInput | ITodoInput | ICalloutInput | IDividerInput | IQuoteInput | IToggleInput | IBulletedListInput | INumberedListInput | ISubSubHeaderInput | ISubHeaderInput | IHeaderInput | ITextInput;
// Advanced block input
export interface ITOCInput extends IInput {
  type: 'table_of_contents',
  format?: ITOC["format"],
  properties?: ITOC["properties"]
}

export interface IEquationInput extends IInput {
  type: 'equation',
  properties: IEquation["properties"],
  format?: IEquation["format"]
}

export interface IFactoryInput extends IInput {
  type: 'factory',
  properties: IFactory["properties"],
  format?: IFactory["format"],
  contents?: TBlockInput[]
}

export interface IBreadcrumbInput extends IInput {
  type: 'breadcrumb',
  properties?: IBreadcrumb["properties"],
  format?: IBreadcrumb["format"],
}

export type TAdvancedBlockInput = IBreadcrumbInput | IFactoryInput | IEquationInput | ITOCInput;

// Embed block input
export interface IEmbedInput extends IInput {
  type: "embed",
  properties: IEmbed["properties"],
  format?: IEmbed["format"],
}

export interface IDriveInput extends IInput {
  type: 'drive',
  properties?: IDrive["properties"],
  format?: IDrive["format"],
  file_id: string
}

export interface ITweetInput extends IInput {
  type: 'tweet',
  properties: ITweet["properties"],
  format?: ITweet["format"],
}

export interface ICodepenInput extends IInput {
  type: 'codepen',
  properties: ICodepen["properties"],
  format?: ICodepen["format"],
}

export interface IMapsInput extends IInput {
  type: 'maps',
  properties: IMaps["properties"],
  format?: IMaps["format"],
}

export interface IGistInput extends IInput {
  type: 'gist',
  properties: IGist["properties"],
  format?: IGist["format"],
}

export interface IFigmaInput extends IInput {
  type: 'figma',
  properties: IFigma["properties"],
  format?: IFigma["format"],
}

export type TEmbedBlockInput = IEmbedInput | IFigmaInput | IMapsInput | ICodepenInput | IDriveInput | IGistInput | ITweetInput;

export type TBlockInput = TMediaBlockInput | TBasicBlockInput | TAdvancedBlockInput | TEmbedBlockInput | TCollectionBlockInput | IColumnListInput;

export type RepositionParams = {
  id: string,
  position: "Before" | "After"
} | number | undefined;

export const TSpaceUpdateKeys = ["name", "icon", "disable_public_access", "disable_guests", "disable_move_to_space", "disable_export", "domain", "invite_link_enabled", "beta_enabled"] as const;
export const TCollectionUpdateKeys = ["name", "icon", "description"] as const;
export const TNotionUserUpdateKeys = ['family_name', 'given_name', 'profile_photo'] as const;
export const TSpaceViewUpdateKeys = ['notify_desktop', 'notify_email', 'notify_mobile', 'joined', 'created_getting_started'] as const;
export const TUserSettingsUpdateKeys = ['start_day_of_week', 'time_zone', 'locale', 'preferred_locale', 'preferred_locale_origin'] as const;

export type ISpaceUpdateInput = Partial<Pick<ISpace, ElementType<typeof TSpaceUpdateKeys>>>;
export type ISpaceCreateInput = Partial<Pick<ISpace, Exclude<ElementType<typeof TSpaceUpdateKeys>, "name" | "domain">>> & {name: string, contents: TBlockCreateInput[]};

export type ICollectionUpdateInput = Partial<Pick<ICollection, ElementType<typeof TCollectionUpdateKeys>>>;

export type INotionUserUpdateInput = Partial<Pick<INotionUser, ElementType<typeof TNotionUserUpdateKeys>>>;

export type ISpaceViewUpdateInput = Partial<Pick<ISpaceView, ElementType<typeof TSpaceViewUpdateKeys>>>;

export type IUserSettingsUpdateInput = Partial<Pick<IUserSettingsSettings, ElementType<typeof TUserSettingsUpdateKeys>>>;

export type TBlockCreateInput = TBlockInput & {
  position?: RepositionParams
}