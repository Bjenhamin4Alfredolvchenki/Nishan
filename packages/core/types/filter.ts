import { TSchemaUnitType, IViewFilterData, TViewGroupFilterOperator } from '@nishans/types';

export interface ViewFilterCreateInput<T extends TSchemaUnitType> {
	filter: {
		operator: IViewFilterData<T>['operator'];
		type: IViewFilterData<T>['type'];
		value: IViewFilterData<T>['value'];
		filter_operator?: TViewGroupFilterOperator;
		children?: ViewFilterCreateInput<TSchemaUnitType>[];
	};
	name: string;
	type: T;
	position?: number;
}

export type TViewFilterCreateInput =
	| ViewFilterCreateInput<'checkbox'>
	| ViewFilterCreateInput<'text'>
	| ViewFilterCreateInput<'number'>
	| ViewFilterCreateInput<'select'>
	| ViewFilterCreateInput<'multi_select'>
	| ViewFilterCreateInput<'title'>
	| ViewFilterCreateInput<'date'>
	| ViewFilterCreateInput<'person'>
	| ViewFilterCreateInput<'file'>
	| ViewFilterCreateInput<'url'>
	| ViewFilterCreateInput<'email'>
	| ViewFilterCreateInput<'phone_number'>
	| ViewFilterCreateInput<'formula'>
	| ViewFilterCreateInput<'relation'>
	| ViewFilterCreateInput<'rollup'>
	| ViewFilterCreateInput<'created_time'>
	| ViewFilterCreateInput<'created_by'>
	| ViewFilterCreateInput<'last_edited_time'>
	| ViewFilterCreateInput<'last_edited_by'>;
