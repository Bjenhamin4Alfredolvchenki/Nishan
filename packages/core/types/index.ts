export * from './block';
export * from './maps';
export * from './nishan';
