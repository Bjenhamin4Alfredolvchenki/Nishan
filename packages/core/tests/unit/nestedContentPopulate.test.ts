import { ICache } from '@nishans/endpoints';
import { IHeader, IOperation } from '@nishans/types';
import deepEqual from 'deep-equal';
import { v4 } from 'uuid';
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';
import colors from 'colors';

import {
	appendChildToParent,
	Block,
	CollectionView,
	CollectionViewPage,
	fetchAndCacheData,
	IHeaderInput,
	nestedContentPopulate,
	Page
} from '../../src';

axios.defaults.baseURL = 'https://www.notion.so/api/v3';
const mock = new MockAdapter(axios);

const space_id = '322c0b0c-5fb4-4fdc-97eb-84e7142b2a80',
	user_id = '72e85506-b758-481a-92b1-73984a903002',
	shard_id = 731776;

const constructDefaultCache = () => {
	return {
		block: new Map([ [ 'page1_id', {} ], [ 'page2_id', { content: [] } ] ]),
		collection: new Map([ [ 'collection1_id', {} ] ]),
		space: new Map([ [ space_id, {} ] ]),
		collection_view: new Map(),
		notion_user: new Map(),
		space_view: new Map(),
		user_root: new Map(),
		user_settings: new Map()
	} as ICache;
};

const metadata = {
	created_time: expect.any(Number),
	created_by_id: user_id,
	created_by_table: 'notion_user',
	last_edited_time: expect.any(Number),
	last_edited_by_table: 'notion_user',
	last_edited_by_id: user_id,
	space_id,
	shard_id,
	version: 0
};

describe('fetchAndCacheData', () => {
	it('exist in cache', async () => {
		const data = await fetchAndCacheData(
			'block',
			'id',
			{
				block: new Map([ [ 'id', { data: 'data' } ] ])
			} as any,
			'token'
		);
		expect(deepEqual(data, { data: 'data' })).toBe(true);
	});

	it.only('doesnt exist in cache', async () => {
		const console_log_spy = jest.spyOn(console, 'log');
		mock.onPost(`/syncRecordValues`).replyOnce(200, {
			recordMap: {
				block: {
					id: {
						role: 'editor',
						value: { data: 'data' }
					}
				}
			}
		});
		const data = await fetchAndCacheData(
			'block',
			'id',
			{
				block: new Map()
			} as any,
			'token'
		);
		expect(console_log_spy).toHaveBeenCalledTimes(1);
		expect(console_log_spy).toHaveBeenCalledWith(colors.yellow.bold(`block:id doesnot exist in the cache`));

		expect(deepEqual(data, { data: 'data' })).toBe(true);
	});
});

describe('appendChildToParent', () => {
	describe(`type=block`, () => {
		it(`path exists`, async () => {
			const stack: IOperation[] = [],
				cache: ICache = {
					block: new Map([ [ 'parent_id', { content: [] } ] ])
				} as any;
			await appendChildToParent('block', 'parent_id', 'child_id', cache, stack, 'token');

			expect(
				deepEqual(stack, [
					{
						table: 'block',
						command: 'listAfter',
						id: 'parent_id',
						args: {
							after: '',
							id: 'child_id'
						},
						path: [ 'content' ]
					}
				])
			).toBe(true);

			expect(deepEqual(cache.block.get('parent_id'), { content: [ 'child_id' ] })).toBe(true);
		});

		it(`path doesnt exists`, async () => {
			const stack: IOperation[] = [],
				cache: ICache = {
					block: new Map([ [ 'parent_id', {} ] ])
				} as any;
			await appendChildToParent('block', 'parent_id', 'child_id', cache, stack, 'token');

			expect(
				deepEqual(stack, [
					{
						table: 'block',
						command: 'listAfter',
						id: 'parent_id',
						args: {
							after: '',
							id: 'child_id'
						},
						path: [ 'content' ]
					}
				])
			).toBe(true);

			expect(deepEqual(cache.block.get('parent_id'), { content: [ 'child_id' ] })).toBe(true);
		});
	});
});

describe('nestedContentPopulate', () => {
	// * Assert the loggers
	// * Assert the returned block map
	// * Assert the cache
	// * Assert the stack

	describe('type=page', () => {
		it(`is_template=false,is_private=true,contents=[{}],parent=space`, async () => {
			const page_id = '48a3d30c-c259-4047-986f-74d9a6cb1305',
				header_id = '48a3d30c-c259-4047-986f-74d9a6cb1304';

			const cache = constructDefaultCache(),
				stack: IOperation[] = [];

			const logger_spy = jest.fn();

			const block_map = await nestedContentPopulate(
				[
					{
						type: 'page',
						properties: { title: [ [ 'Page' ] ] },
						format: {
							page_font: 'Monaco'
						},
						isPrivate: true,
						id: page_id,
						contents: [
							{
								type: 'header',
								properties: {
									title: [ [ 'Header' ] ]
								},
								format: {
									block_color: 'blue'
								},
								id: header_id
							}
						]
					}
				],
				space_id,
				'space',
				{
					cache,
					interval: 0,
					logger: logger_spy,
					shard_id,
					space_id,
					stack,
					token: 'token',
					user_id
				}
			);

			const header_snapshot = {
					id: header_id,
					type: 'header',
					properties: {
						title: [ [ 'Header' ] ]
					},
					format: {
						block_color: 'blue'
					},
					parent_id: page_id,
					parent_table: 'block',
					...metadata
				},
				page_snapshot = {
					id: expect.any(String),
					type: 'page',
					properties: {
						title: [ [ 'Page' ] ]
					},
					format: {
						page_font: 'Monaco'
					},
					parent_id: space_id,
					parent_table: 'space',
					alive: true,
					permissions: [
						{
							type: 'user_permission',
							role: 'editor',
							user_id
						}
					],
					created_time: expect.any(Number),
					created_by_id: user_id,
					created_by_table: 'notion_user',
					last_edited_time: expect.any(Number),
					last_edited_by_table: 'notion_user',
					last_edited_by_id: user_id,
					space_id,
					shard_id,
					version: 0
				};

			expect(logger_spy).toHaveBeenCalledTimes(2);
			expect(logger_spy).toHaveBeenNthCalledWith(1, 'CREATE', 'block', header_id);
			expect(logger_spy).toHaveBeenNthCalledWith(2, 'CREATE', 'block', page_id);

			expect((block_map.page.get('Page') as Page).getCachedData()).toMatchSnapshot({
				...page_snapshot,
				content: [ header_id ]
			});

			expect((block_map.header.get(header_id) as Block<IHeader, IHeaderInput>).getCachedData()).toMatchSnapshot(
				header_snapshot
			);

			expect(stack).toMatchSnapshot([
				{
					table: 'block',
					command: 'update',
					path: [],
					id: page_id,
					args: page_snapshot
				},
				{
					table: 'block',
					command: 'update',
					path: [],
					id: header_id,
					args: header_snapshot
				},
				{
					path: [ 'content' ],
					table: 'block',
					command: 'listAfter',
					args: { after: '', id: header_id },
					id: page_id
				},
				{
					path: [ 'pages' ],
					table: 'space',
					command: 'listAfter',
					args: { after: '', id: page_id },
					id: space_id
				}
			]);
			expect(
				deepEqual(cache.space.get(space_id), {
					pages: [ page_id ]
				})
			).toBe(true);
		});
	});

	describe('type=collection_block', () => {
		it(`type=collection_view_page,rows=[]`, async () => {
			const cache = constructDefaultCache(),
				stack: IOperation[] = [];

			const logger_spy = jest.fn();

			const block_map = await nestedContentPopulate(
				[
					{
						type: 'collection_view_page',
						name: [ [ 'Collection Name' ] ],
						schema: [
							{
								type: 'title',
								name: 'Title'
							}
						],
						views: [
							{
								type: 'table',
								name: 'Table',
								schema_units: [
									{
										type: 'title',
										name: 'Title'
									}
								]
							}
						]
					}
				],
				space_id,
				'space',
				{
					cache,
					shard_id,
					space_id,
					interval: 0,
					logger: logger_spy,
					stack,
					token: 'token',
					user_id
				}
			);

			const collection_view_page = (block_map.collection_view_page.get(
				'Collection Name'
			) as CollectionViewPage).getCachedData();

			expect(logger_spy).toHaveBeenCalledTimes(3);
			expect(logger_spy).toHaveBeenNthCalledWith(1, 'CREATE', 'collection_view', collection_view_page.view_ids[0]);
			expect(logger_spy).toHaveBeenNthCalledWith(2, 'CREATE', 'collection', collection_view_page.collection_id);
			expect(logger_spy).toHaveBeenNthCalledWith(3, 'CREATE', 'block', collection_view_page.id);

			const collection_view_page_snapshot = {
				id: expect.any(String),
				type: 'collection_view_page',
				collection_id: expect.any(String),
				view_ids: [ expect.any(String) ],
				parent_id: space_id,
				parent_table: 'space',
				permissions: [
					{
						type: 'space_permission',
						role: 'editor',
						user_id
					}
				],
				...metadata
			};
			expect(collection_view_page).toMatchSnapshot(collection_view_page_snapshot);

			expect(cache.block.get(collection_view_page.id)).toBeTruthy();
			expect(cache.collection.get(collection_view_page.collection_id)).toBeTruthy();
			expect(cache.collection_view.get(collection_view_page.view_ids[0])).toBeTruthy();
			expect(stack.length).toBe(4);

			expect(stack[2]).toMatchSnapshot({
				path: [],
				table: 'block',
				command: 'update',
				args: collection_view_page_snapshot,
				id: expect.any(String)
			});
		});

		it(`type=collection_view,rows=[{}]`, async () => {
			const cache = constructDefaultCache(),
				stack: IOperation[] = [],
				row_one_id = v4();

			const logger_spy = jest.fn();

			const block_map = await nestedContentPopulate(
				[
					{
						type: 'collection_view',
						name: [ [ 'Collection Name' ] ],
						schema: [
							{
								type: 'title',
								name: 'Title'
							}
						],
						views: [
							{
								type: 'table',
								name: 'Table',
								schema_units: [
									{
										type: 'title',
										name: 'Title'
									}
								]
							}
						],
						rows: [
							{
								type: 'page',
								id: row_one_id,
								properties: {
									title: [ [ 'Row one' ] ]
								}
							}
						]
					}
				],
				'page1_id',
				'block',
				{
					cache,
					shard_id,
					space_id,
					interval: 0,
					logger: logger_spy,
					stack,
					token: 'token',
					user_id
				}
			);

			const collection_view = (block_map.collection_view.get('Collection Name') as CollectionView).getCachedData();

			expect(logger_spy).toHaveBeenCalledTimes(4);
			expect(logger_spy).toHaveBeenNthCalledWith(1, 'CREATE', 'collection_view', collection_view.view_ids[0]);
			expect(logger_spy).toHaveBeenNthCalledWith(2, 'CREATE', 'collection', collection_view.collection_id);
			expect(logger_spy).toHaveBeenNthCalledWith(3, 'CREATE', 'block', row_one_id);
			expect(logger_spy).toHaveBeenNthCalledWith(4, 'CREATE', 'block', collection_view.id);

			const collection_view_snapshot = {
				id: expect.any(String),
				type: 'collection_view',
				collection_id: expect.any(String),
				view_ids: [ expect.any(String) ],
				parent_id: 'page1_id',
				parent_table: 'block',
				...metadata
			};
			expect(collection_view).toMatchSnapshot(collection_view_snapshot);

			expect(cache.block.get(collection_view.id)).toBeTruthy();
			expect(cache.block.get(row_one_id)).toBeTruthy();
			expect(cache.collection.get(collection_view.collection_id)).toBeTruthy();
			expect(cache.collection_view.get(collection_view.view_ids[0])).toBeTruthy();

			expect(stack.length).toBe(5);

			expect(stack[2]).toMatchSnapshot({
				path: [],
				table: 'block',
				command: 'update',
				args: collection_view_snapshot,
				id: expect.any(String)
			});
		});
	});
});
