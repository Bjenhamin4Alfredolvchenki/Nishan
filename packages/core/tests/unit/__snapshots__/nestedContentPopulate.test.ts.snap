// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nestedContentPopulate type=collection_block type=collection_view,rows=[{}] 1`] = `
Object {
  "alive": true,
  "collection_id": Any<String>,
  "created_by_id": "72e85506-b758-481a-92b1-73984a903002",
  "created_by_table": "notion_user",
  "created_time": Any<Number>,
  "id": Any<String>,
  "last_edited_by_id": "72e85506-b758-481a-92b1-73984a903002",
  "last_edited_by_table": "notion_user",
  "last_edited_time": Any<Number>,
  "parent_id": "page1_id",
  "parent_table": "block",
  "shard_id": 731776,
  "space_id": "322c0b0c-5fb4-4fdc-97eb-84e7142b2a80",
  "type": "collection_view",
  "version": 0,
  "view_ids": Array [
    Any<String>,
  ],
}
`;

exports[`nestedContentPopulate type=collection_block type=collection_view,rows=[{}] 2`] = `
Object {
  "args": Object {
    "alive": true,
    "collection_id": Any<String>,
    "created_by_id": "72e85506-b758-481a-92b1-73984a903002",
    "created_by_table": "notion_user",
    "created_time": Any<Number>,
    "id": Any<String>,
    "last_edited_by_id": "72e85506-b758-481a-92b1-73984a903002",
    "last_edited_by_table": "notion_user",
    "last_edited_time": Any<Number>,
    "parent_id": "page1_id",
    "parent_table": "block",
    "shard_id": 731776,
    "space_id": "322c0b0c-5fb4-4fdc-97eb-84e7142b2a80",
    "type": "collection_view",
    "version": 0,
    "view_ids": Array [
      Any<String>,
    ],
  },
  "command": "update",
  "id": Any<String>,
  "path": Array [],
  "table": "block",
}
`;

exports[`nestedContentPopulate type=collection_block type=collection_view_page,rows=[] 1`] = `
Object {
  "alive": true,
  "collection_id": Any<String>,
  "created_by_id": "72e85506-b758-481a-92b1-73984a903002",
  "created_by_table": "notion_user",
  "created_time": Any<Number>,
  "id": Any<String>,
  "last_edited_by_id": "72e85506-b758-481a-92b1-73984a903002",
  "last_edited_by_table": "notion_user",
  "last_edited_time": Any<Number>,
  "parent_id": "322c0b0c-5fb4-4fdc-97eb-84e7142b2a80",
  "parent_table": "space",
  "permissions": Array [
    Object {
      "role": "editor",
      "type": "space_permission",
      "user_id": "72e85506-b758-481a-92b1-73984a903002",
    },
  ],
  "shard_id": 731776,
  "space_id": "322c0b0c-5fb4-4fdc-97eb-84e7142b2a80",
  "type": "collection_view_page",
  "version": 0,
  "view_ids": Array [
    Any<String>,
  ],
}
`;

exports[`nestedContentPopulate type=collection_block type=collection_view_page,rows=[] 2`] = `
Object {
  "args": Object {
    "alive": true,
    "collection_id": Any<String>,
    "created_by_id": "72e85506-b758-481a-92b1-73984a903002",
    "created_by_table": "notion_user",
    "created_time": Any<Number>,
    "id": Any<String>,
    "last_edited_by_id": "72e85506-b758-481a-92b1-73984a903002",
    "last_edited_by_table": "notion_user",
    "last_edited_time": Any<Number>,
    "parent_id": "322c0b0c-5fb4-4fdc-97eb-84e7142b2a80",
    "parent_table": "space",
    "permissions": Array [
      Object {
        "role": "editor",
        "type": "space_permission",
        "user_id": "72e85506-b758-481a-92b1-73984a903002",
      },
    ],
    "shard_id": 731776,
    "space_id": "322c0b0c-5fb4-4fdc-97eb-84e7142b2a80",
    "type": "collection_view_page",
    "version": 0,
    "view_ids": Array [
      Any<String>,
    ],
  },
  "command": "update",
  "id": Any<String>,
  "path": Array [],
  "table": "block",
}
`;

exports[`nestedContentPopulate type=page is_template=false,is_private=true,contents=[{}],parent=space 1`] = `
Object {
  "alive": true,
  "content": Array [
    "48a3d30c-c259-4047-986f-74d9a6cb1304",
  ],
  "created_by_id": "72e85506-b758-481a-92b1-73984a903002",
  "created_by_table": "notion_user",
  "created_time": Any<Number>,
  "format": Object {
    "page_font": "Monaco",
  },
  "id": Any<String>,
  "last_edited_by_id": "72e85506-b758-481a-92b1-73984a903002",
  "last_edited_by_table": "notion_user",
  "last_edited_time": Any<Number>,
  "parent_id": "322c0b0c-5fb4-4fdc-97eb-84e7142b2a80",
  "parent_table": "space",
  "permissions": Array [
    Object {
      "role": "editor",
      "type": "user_permission",
      "user_id": "72e85506-b758-481a-92b1-73984a903002",
    },
  ],
  "properties": Object {
    "title": Array [
      Array [
        "Page",
      ],
    ],
  },
  "shard_id": 731776,
  "space_id": "322c0b0c-5fb4-4fdc-97eb-84e7142b2a80",
  "type": "page",
  "version": 0,
}
`;

exports[`nestedContentPopulate type=page is_template=false,is_private=true,contents=[{}],parent=space 2`] = `
Object {
  "alive": true,
  "created_by_id": "72e85506-b758-481a-92b1-73984a903002",
  "created_by_table": "notion_user",
  "created_time": Any<Number>,
  "format": Object {
    "block_color": "blue",
  },
  "id": "48a3d30c-c259-4047-986f-74d9a6cb1304",
  "last_edited_by_id": "72e85506-b758-481a-92b1-73984a903002",
  "last_edited_by_table": "notion_user",
  "last_edited_time": Any<Number>,
  "parent_id": "48a3d30c-c259-4047-986f-74d9a6cb1305",
  "parent_table": "block",
  "properties": Object {
    "title": Array [
      Array [
        "Header",
      ],
    ],
  },
  "shard_id": 731776,
  "space_id": "322c0b0c-5fb4-4fdc-97eb-84e7142b2a80",
  "type": "header",
  "version": 0,
}
`;

exports[`nestedContentPopulate type=page is_template=false,is_private=true,contents=[{}],parent=space 3`] = `
Array [
  Object {
    "args": Object {
      "alive": true,
      "content": Array [],
      "created_by_id": "72e85506-b758-481a-92b1-73984a903002",
      "created_by_table": "notion_user",
      "created_time": Any<Number>,
      "format": Object {
        "page_font": "Monaco",
      },
      "id": Any<String>,
      "last_edited_by_id": "72e85506-b758-481a-92b1-73984a903002",
      "last_edited_by_table": "notion_user",
      "last_edited_time": Any<Number>,
      "parent_id": "322c0b0c-5fb4-4fdc-97eb-84e7142b2a80",
      "parent_table": "space",
      "permissions": Array [
        Object {
          "role": "editor",
          "type": "user_permission",
          "user_id": "72e85506-b758-481a-92b1-73984a903002",
        },
      ],
      "properties": Object {
        "title": Array [
          Array [
            "Page",
          ],
        ],
      },
      "shard_id": 731776,
      "space_id": "322c0b0c-5fb4-4fdc-97eb-84e7142b2a80",
      "type": "page",
      "version": 0,
    },
    "command": "update",
    "id": "48a3d30c-c259-4047-986f-74d9a6cb1305",
    "path": Array [],
    "table": "block",
  },
  Object {
    "args": Object {
      "alive": true,
      "created_by_id": "72e85506-b758-481a-92b1-73984a903002",
      "created_by_table": "notion_user",
      "created_time": Any<Number>,
      "format": Object {
        "block_color": "blue",
      },
      "id": "48a3d30c-c259-4047-986f-74d9a6cb1304",
      "last_edited_by_id": "72e85506-b758-481a-92b1-73984a903002",
      "last_edited_by_table": "notion_user",
      "last_edited_time": Any<Number>,
      "parent_id": "48a3d30c-c259-4047-986f-74d9a6cb1305",
      "parent_table": "block",
      "properties": Object {
        "title": Array [
          Array [
            "Header",
          ],
        ],
      },
      "shard_id": 731776,
      "space_id": "322c0b0c-5fb4-4fdc-97eb-84e7142b2a80",
      "type": "header",
      "version": 0,
    },
    "command": "update",
    "id": "48a3d30c-c259-4047-986f-74d9a6cb1304",
    "path": Array [],
    "table": "block",
  },
  Object {
    "args": Object {
      "after": "",
      "id": "48a3d30c-c259-4047-986f-74d9a6cb1304",
    },
    "command": "listAfter",
    "id": "48a3d30c-c259-4047-986f-74d9a6cb1305",
    "path": Array [
      "content",
    ],
    "table": "block",
  },
  Object {
    "args": Object {
      "after": "",
      "id": "48a3d30c-c259-4047-986f-74d9a6cb1305",
    },
    "command": "listAfter",
    "id": "322c0b0c-5fb4-4fdc-97eb-84e7142b2a80",
    "path": Array [
      "pages",
    ],
    "table": "space",
  },
]
`;
